### PR TITLE
fetch: use sequoia for verifying files

### PR DESCRIPTION
**Pull request by fepitre**
Related to https://github.com/QubesOS/qubes-issues/issues/9012

### CODE REVIEW

File: qubesbuilder/plugins/fetch/scripts/verify-file [hunk 1]
```diff
@@ -114,12 +114,10 @@ elif [ -n "${UNTRUSTED_SIGNATURE_FILE}" ] && [ -n "${PUBKEY_FILE}" ]; then
...
     # Import pubkey
     keyring="$(mktemp -u)"
-    for pubkey_file in "${PUBKEY_FILE[@]}"; do
-        gpg --no-auto-check-trustdb --no-default-keyring --keyring "$keyring" -q --import "${pubkey_file}"
-    done
+    sq keyring merge --output "$keyring" "${PUBKEY_FILE[@]}"
 
     # Verify file
-    gpgv --keyring "$keyring" "${UNTRUSTED_SIGNATURE_FILE}" "${UNTRUSTED_FILE}" 2>/dev/null
+    sqv --keyring "$keyring" "${UNTRUSTED_SIGNATURE_FILE}" "${UNTRUSTED_FILE}"
```
[the remark this answers is not in the document]
Comment (fepitre) [reply qubesbuilder/plugins/fetch/scripts/verify-file:120]: Yes, notably preventing hiding errors.

File: qubesbuilder/plugins/fetch/scripts/get-and-verify-source.py [hunk 2]
```diff
@@ -214,17 +214,26 @@ def main(args):
             print("--> Verifying tags...")
 
         env = {"GNUPGHOME": str(git_keyring_dir)}
-        if not (git_keyring_dir / "trustdb.gpg").exists():
+        trust_root = git_keyring_dir / "pubring.cert.d" / "trust-root"
+        if not trust_root.exists():
             git_keyring_dir.mkdir(parents=True, exist_ok=True)
             git_keyring_dir.chmod(0o700)
+            # We request a list to init the keyring. It looks like it does
+            # not do it on first import, so we just show available keys.
```
[the remark this answers is not in the document]
Comment (fepitre) [reply qubesbuilder/plugins/fetch/scripts/get-and-verify-source.py:222]: Without it:
```
17:10:05,278 [executor:local:/var/tmp/128182026454704d778fec0/builder] output: [Errno 2] No such file or directory: '/var/tmp/128182026454704d778fec0/builder/keyring/pubring.cert.d/trust-root'
Error: Failed to run '['bash', '-c', 'cd /var/tmp/128182026454704d778fec0/builder&&/var/tmp/128182026454704d778fec0/builder/plugins/fetch/scripts/get-and-verify-source.py https://github.com/QubesOS/qubes-core-qrexec.git /var/tmp/128182026454704d778fec0/builder/core-qrexec /var/tmp/128182026454704d778fec0/builder/keyring /var/tmp/128182026454704d778fec0/builder/plugins/fetch/keys --git-branch main --minimum-distinct-maintainers 1 --maintainer 0064428F455451B3EBE78A7F063938BA42CFA724 --maintainer 274E12AB03F2FE293765FC06DA0434BC706E1FCF']' (status=1).
```

File: README.md [hunk 3]
```diff
@@ -17,13 +17,13 @@ Local, and Qubes executors are available.
 Fedora:
 
 ```bash
-$ sudo dnf install python3-packaging python3-click python3-lxml createrepo_c devscripts gpg python3-pyyaml rpm docker python3-docker podman python3-podman reprepro python3-pathspec rpm-sign rb_libtorrent-examples openssl tree mock python3-jinja2-cli pacman m4 asciidoc rsync
+$ sudo dnf install python3-packaging python3-click python3-lxml createrepo_c devscripts gpg python3-pyyaml rpm docker python3-docker podman python3-podman reprepro python3-pathspec rpm-sign rb_libtorrent-examples openssl tree mock python3-jinja2-cli pacman m4 asciidoc rsync sequoia-sq sequoia-sqv
```
[the remark this answers is not in the document]
Comment (fepitre) [reply README.md:20]: @marmarek I think sequoia-chameleon-gnupg is not available in Debian.

[the remark this answers is not in the document]
Comment (marmarek) [reply README.md:20]: This means this change won't work at all on Debian, which is not good. In this case, either limit this PR to just tarball (sqv is on Debian, right?), or use some fallback so Debian will keep using plain gnupg. I'd go with the former, but if you want to implement the fallback, I'm okay with that too.
And then, lets return to this topic when Debian 13 will be released.

**Review by marmarek — APPROVED** [review]

**Comment by marmarek** [comment]
The tag verification via `gpg-sq` fails. After [adding some debugging](https://gitlab.com/QubesOS/qubes-builderv2/-/jobs/6508505890), I see `[GNUPG:] TRUST_UNDEFINED 0 pgp` in the output. The test passes on wasabi runner (I guess gpg-sq not installed there, so the fallback works), but fails on airelle.

**Comment by marmarek** [comment]
@nwalfield I'm pretty sure it's there already: https://gitlab.com/sequoia-pgp/sequoia-chameleon-gnupg/-/issues/34

We have a workaround already: mark all relevant keys as ultimately trusted (we have list of them in the config anyway).

**Comment by nwalfield** [comment]
> @nwalfield I'm pretty sure it's there already: https://gitlab.com/sequoia-pgp/sequoia-chameleon-gnupg/-/issues/34
> 
> We have a workaround already: mark all relevant keys as ultimately trusted (we have list of them in the config anyway).

Thanks for clarifying.  If you mark the keys as ultimately trusted, then you are saying that the certificates are also CAs, so you need to make sure that the keyring is otherwise empty.  Otherwise, if Alice is not considered authorized according to the config file, but Bob is, and he certified Alice, `gpg` will consider the commit to be authenticated.

**Comment by marmarek** [comment]
Yes, this script creates a temporary keyring with just the authorized keys (and the new version marks all of them as ultimately trusted).

**Pull request merged**